File: leetcode/698.划分为k个相等的子集/698.划分为k个相等的子集.py
from typing import List


class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        n = len(nums)
        if k > n:
            return False
        nums_sum = sum(nums)
        # 总值%桶数 == 0 ，可以被整除才能进行元素和查找
        if nums_sum % k != 0:
            return False
        # 每个桶的和
        target = nums_sum / k
        # 标记元素使用状态，避免重复
        used = [False for i in range(n)]

        def back_track(bucket_num, bucket_sum, start):
            # base case 桶数为0, 所有桶都装满了
            if bucket_num == 0:
                return True
            # 装满了当前桶，递归穷举下一个桶的选择 让下一个桶从 nums[0] 开始选数字
            if bucket_sum == target:
                return back_track(bucket_num - 1, 0, 0)

            # 从start开始向后查询是否有有效的nums[1]
            for i in range(start, len(nums)):
                # nums[i]已经被使用
                if used[i]:
                    continue
                # 装不下
                if nums[i] + bucket_sum > target:
                    continue
                # 做选择，将nums[i]装入
                used[i] = True
                bucket_sum += nums[i]
                # 递归到下一层
                res = back_track(bucket_num, bucket_sum, i + 1)
                if res:
                    return True
                # 撤销选择
                used[i] = False
                bucket_sum -= nums[i]

            # 穷举了所有数字，都无法达到目标
            return False

        return back_track(k, 0, 0)


if __name__ == '__main__':
    nums = [4, 3, 2, 3, 5, 2, 1]
    k = 4
    result = Solution().canPartitionKSubsets(nums, k)
    print(result)

File: leetcode/698.划分为k个相等的子集/698.划分为k个相等的子集.go
package main

func canPartitionKSubsets(nums []int, k int) bool {
	// 桶数超过数组长度
	n := len(nums)
	if k > n {
		return false
	}
	// 数组和 平均分 桶
	sum := 0
	for _, num := range nums {
		sum += num
	}
	if sum%k != 0 {
		return false
	}
	// 平均分桶后每桶的 总和
	target := sum / k

	// 值是否被使用
	used := make([]bool, n)
	for i := 0; i < n; i++ {
		used[i] = false
	}
	var backTrack func(bucketNum int, bucketSum int, start int) bool
	backTrack = func(bucketNum int, bucketSum int, start int) bool {
		// base case 回溯结束，所有桶都满了
		if bucketNum == 0 {
			return true
		}
		// 当前桶 已满，从头开始找下个桶
		if bucketSum == target {
			return backTrack(bucketNum-1, 0, 0)
		}
		// 遍历选择
		for i := start; i < n; i++ {
			// 是否已经被使用
			if used[i] {
				continue
			}
			// 是否装不下
			if nums[i]+bucketSum > target {
				continue
			}
			// 做选择
			used[i] = true
			bucketSum += nums[i]
			// 递归下一级, 前进一步
			res := backTrack(bucketNum, bucketSum, i+1)
			if res == true {
				return true
			}
			// 撤销选择
			used[i] = false
			bucketSum -= nums[i]
		}
		// 没有合适就返回false
		return false
	}

	return backTrack(k, 0, 0)
}

func main() {
	nums := []int{4, 3, 2, 3, 5, 2, 1}
	k := 4
	result := canPartitionKSubsets(nums, k)
	println(result)
}
